src/storage/kv/rocksdb_engine.rs

```diff
@@ -290,13 +290,26 @@ impl Engine for RocksEngine {
         write_modifies(&self.engines.kv, modifies)
     }
 
-    fn async_write(&self, _: &Context, batch: WriteData, cb: Callback<()>) -> Result<()> {
+    fn async_write(&self, ctx: &Context, batch: WriteData, cb: Callback<()>) -> Result<()> {
+        self.async_write_ext(ctx, batch, cb, None)
+    }
+
+    fn async_write_ext(
+        &self,
+        _: &Context,
+        batch: WriteData,
+        cb: Callback<()>,
+        proposed_cb: Option<ExtCallback>,
+    ) -> Result<()> {
         fail_point!("rockskv_async_write", |_| Err(box_err!("write failed")));
 
         if batch.modifies.is_empty() {
             return Err(Error::from(ErrorInner::EmptyRequest));
         }
         box_try!(self.sched.schedule(Task::Write(batch.modifies, cb)));
+        if let Some(cb) = proposed_cb {
```

Comment:
Hmmm. I change this order now. Actually, we can guarantee that the `proposed_cb` must be called before `cb` in `raftkv` engine but this order may still be reversed in `txn::scheduler`(I am not sure, @youjiali1995 told me that)

Reply:
`proposed_cb` and `write_cb` are invoked in order but what they do are not ordered since these callbacks schedule a new task to the scheduler pool which is undetermined.

Reply:
Yes. It has no difference for pipelined pessimistic lock. I just mention it for possible future use when the order counts.